### PR TITLE
remove regex's, fix entry module

### PR DESCRIPTION
**Pull request by jbellenger**
This is builds on [PR 51: add support for string module ids](https://github.com/rxaviers/globalize-webpack-plugin/pull/51). A simpler delta between this and PR 51 is [here](https://github.com/jbellenger/globalize-webpack-plugin/compare/master...jbellenger:entry-ordering)

**Background**
At twitter, we want to adopt webpack2 so that we can leverage the HashedModuleIds plugin. This will create stable hashes for our modules and chunks, improving caching for our users.

**Problem**
The present method of rendering localized chunks does not behave well when using string module ids and webpack2. ProductionModePlugin puts the globalize-compiled-data into the first module of a chunk, assuming that the first module is always executed as the entry module. This is not the case in webpack2, where the id of the entry module is an argument to webpackJsonp and rendered into the chunk.

**Proposal**
Replace the regex module rewriting that's done post-render with something more webpacky that can be done pre-render. This is essentially the same behavior as what is done currently, but using webpack primitives instead of strings. Using these webpack primitives, we can identify the chunks entry module and ensure that modules are rewritten correctly.

**TODOs**
- [x] test with webpack1
- [x] test in production
- [x] address feedback

### CODE REVIEW

File: ProductionModePlugin.js [hunk 1]
```diff
@@ -191,66 +190,93 @@ ProductionModePlugin.prototype.apply = function(compiler) {
...
-          var isFirst = true;
-          // Match either webpack 1.x or webpack 2.x
-          var source = compilation.assets[file].source().replace(/\n\/\*\*\*\/ (\(?function)\(module, exports(, __webpack_require__)?\) {[\s\S]*?(\n\/\*\*\*\/ })/g, function(garbage1, fnHead, garbage2, fnTail) {
-            var fnContent;
+      // rewrite the modules in the localized chunks:
+      //   - entry module will contain the compiled formatters and parsers
+      //   - non-entry modules will be rewritten to export globalize
+      this.chunks
+        .filter((chunk) => /globalize-compiled-data/.test(chunk.name))
+        .forEach((chunk) => {
```
Comment (rxaviers) [inline ProductionModePlugin.js]: Please, update `package.json` node engine accordingly for such support, i.e., arrow functions, const, etc.

File: test/ProductionModePlugin.js [hunk 2]
```diff
@@ -1,37 +1,55 @@
...
 var GlobalizePlugin = require("../index");
-var PathChunkPlugin = require("path-chunk-webpack-plugin");
+var expect = require("chai").expect;
+var fs = require("fs");
+var mkdirp = require("mkdirp");
 var path = require("path");
-var webpack = require("webpack");
 var rimraf = require("rimraf");
-var fs = require("fs");
+var webpack = require("webpack");
```
Comment (rxaviers) [inline test/ProductionModePlugin.js]: nitpick: let's sort this list :)

**Comment by rxaviers** [comment]
Thanks @jbellenger! Please, will that still work on webpack1? If it won't, let's simply remove the hacks to support both simultaneously and we tag it to lend as a next major.

File: package.json [hunk 3]
```diff
@@ -38,10 +42,11 @@
     "eslint": "^3.19.0",
     "eslint-config-defaults": "^9.0.0",
     "globalize": "^1.2.3",
+    "mkdirp": "^0.5.1",
     "mocha": "^3.3.0",
     "path-chunk-webpack-plugin": "^1.2.0",
     "rimraf": "^2.6.1",
-    "webpack": "^1.15.0"
+    "webpack": "^2.5.1"
```
Comment (rxaviers) [inline package.json]: :+1:

File: ProductionModePlugin.js [hunk 4]
```diff
@@ -191,66 +190,93 @@ ProductionModePlugin.prototype.apply = function(compiler) {
...
-            return "\n/***/ " + fnHead + "(module, exports, __webpack_require__) {\n" + fnContent + fnTail;
+            // The `module` object in scope here is in each locale chunk, and
+            // any modifications we make will be rendered into every locale
+            // chunk. Create a new module to contain the locale-specific source
+            // modifications we've made.
+            const newModule = new RawModule(fnContent);
+            newModule.context = module.context;
+            newModule.id = module.id;
+            newModule.dependencies = module.dependencies;
+            return newModule;
```
Comment (rxaviers) [inline ProductionModePlugin.js]: This looks wonderful

File: ProductionModePlugin.js [hunk 5]
```diff
@@ -191,66 +190,93 @@ ProductionModePlugin.prototype.apply = function(compiler) {
...
+            let fnContent;
+            if (module === chunk.entryModule) {
+              var locale = chunk.name.replace("globalize-compiled-data-", "");
               fnContent = globalizeCompilerHelper.compile(locale)
                 .replace("typeof define === \"function\" && define.amd", "false")
                 .replace(/require\("([^)]+)"\)/g, function(garbage, moduleName) {
                   return "__webpack_require__(" + globalizeModuleIdsMap[moduleName] + ")";
                 });
-
-            // Define all other individual globalize compiled data as a simple exports to Globalize.
```
Comment (rxaviers) [inline ProductionModePlugin.js]: Should we leave this comment? Otherwise, the `else` code below would be pretty confusing...

**Comment by rxaviers** [comment]
This looks great!

**Comment by jbellenger** [comment]
Thanks for the comments, Rafael! I will:
- address your feedback
- test this with webpack1
- test this in production at twitter

I'll run through these and update this PR next week.

File: test/ProductionModePlugin.js (hunk 2 above)
Comment (jbellenger) [inline test/ProductionModePlugin.js]: Done. Added a rule for [sort-requires](https://www.npmjs.com/package/eslint-plugin-sort-requires) to this PR

File: ProductionModePlugin.js (hunk 5 above)
Comment (jbellenger) [inline ProductionModePlugin.js]: Comment restored.

**Comment by jbellenger** [comment]
**webpack 1 dropped**
I took a stab at running this with webpack1. It mostly works, though locale chunks generate this error on load:
```
Uncaught ReferenceError: __webpack_require__ is not defined
    at Object.<anonymous> (en.6860147fcfea9f96.js:1)
    at a (manifest.0f58aea068d86a70.js:1)
    at window.webpackJsonp (manifest.0f58aea068d86a70.js:1)
    at en.6860147fcfea9f96.js:1
```
I think this could probably be fixed, but for the sake of simplicity I've just removed support for webpack1 and bumped the version to 0.4.0

It looks like the [example project](https://github.com/globalizejs/globalize/tree/master/examples/app-npm-webpack) referenced in README.md might need to be updated.

**Production testing looks good**
I deployed this to 1% of twitter mobile web traffic for 3 hours and saw no increase in client errors. I'm planning on moving all of mobile web traffic to webpack2 (and this code) later this week.

As far as I know, this is ready to merge.

File: ProductionModePlugin.js (hunk 1 above)
Comment (jbellenger) [inline ProductionModePlugin.js]: To reduce noise, I reverted all let/const/arrowfunctions. I'm happy to add them back in as a followup.

**Comment by rxaviers** [comment]
Merged via 23b7dbe1e01aadaf3fa55a6881db9cd6af7ae290

**Comment by rxaviers** [comment]
Thank you!!

**Comment by rxaviers** [comment]
Released `1.0.0-alpha.0` pre-release (webpack-2 only support).

**Pull request closed without merging**